### PR TITLE
Make installer process safer

### LINKED ISSUE #16122: Many users reported that the upgrade process failed because files could not be replaced and caused NVDA corruption.
### Steps to reproduce:
Three patch versions have been pushed recently, which seems to have made this problem more obvious.
There are many reports from users that the upgrade cannot go smoothly because the files cannot be replaced, and even after the user clicks "retry" several times, the upgrade still fails and NVDA becomes corrupted.
After restarting the computer NVDA goes completely silent and they have to ask someone else or reinstall NVDA with the help of Narrator.

### Actual behavior:
The upgrade process failed because some files could not be replaced.
### Expected behavior:
Can be upgraded smoothly.
### NVDA logs, crash dumps and other attachments:
Due to the severity of the issue, users had to reinstall NVDA after their NVDA was compromised, which also resulted in log loss.
Since I have not encountered this problem myself, there is no log for the time being.
#### NVDA installed/portable/running from source:
Installed
#### NVDA version:
2023.3,2023.3.1,2023.3.2
#### Windows version:
Windows10 and Windows11
#### Name and version of other software in use when reproducing the issue:
None
#### Other information about your system:
None
#### Does the issue still occur after restarting your computer?
Some users were able to upgrade successfully after restarting their computer, but others had their NVDA corrupted and unable to boot.
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
I've experienced this myself with other versions of NVDA, but it's become an issue too frequently lately.
#### If NVDA add-ons are disabled, is your problem still occurring?
Unknown at the moment
#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
Unknown at the moment

### PR DESCRIPTION
**Pull request by seanbudd**
### Link to issue number:
Fixes #16122

### Summary of the issue:
NVDA process sometimes fails to quit correctly (see #16123).
When running the installer, this causes the installer to fail partway through.
If the installer fails partway through, the installation may end up in an unrecoverable state.

For certain files, NVDA requests windows to move or delete these on reboot, if the move/delete fails initially.
However, if the Windows API request fails, NVDA does not handle the failure the same way as it handles other failures - raising a retryable failure and reverting the change if possible.

### Description of user facing changes
The NVDA installation process should be more resilient if the previous process is still running.

### Description of development approach

NVDA attempts to remove all known exes in a batch as the first operation in the installation. If any known NVDA exe is running, the installation is aborted with a retry able failure and all exes returned to their original state. This means the user can attempt the installation again - i.e. waiting a bit longer for the process to hopefully exit.

If NVDA fails to request windows to remove certain files on restart, this is handled the same as if the files are failed to be removed.

More logging has been added and improved.
### Testing strategy:
- [x] Tested manually triggering a retriable failure to ensure the handling is as expected.
- [ ] Manual testing from the community is required to confirm the fix.

### Known issues with pull request:
The installation process should be atomic - i.e. if it fails, the whole operation should be reverted. This is not the case of the current design. In future code to NVDA we can specify more files we want to include in the atomic and reversible operation.

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [ ] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

### CODE REVIEW

File: appx/sconscript [hunk 1]
```diff
@@ -63,7 +63,8 @@ signExec=env['signExec'] if env['certFile'] else None
 # Files from NVDA's distribution that cannot be included in the appx due to policy or security restrictions
 excludedDistFiles=[
 	'nvda_eoaProxy.exe',
-	'nvda_service.exe',
+	# TODO should this be excluded?
+	# 'nvda_dmp.exe',
```
Comment (michaelDCurran) [inline appx/sconscript:67]: I don't think this should be excluded. Though perhaps a second exe may need to be specifically mentioned in the manifest... We don't really use this code at the momement, and would only really know once we build one again and verify it with Microsoft tools. But it is correct that nvda_service can be removed as we don't create it anymore.

File: source/installer.py [hunk 2]
```diff
@@ -566,10 +567,17 @@ def _deleteKeyAndSubkeys(key, subkey):
...
+def tryRemoveFile(
+		path: str,
+		numRetries: int = 6,
+		retryInterval: float = 0.5,
+		rebootOK: bool = False
+):
 	dirPath=os.path.dirname(path)
 	tempPath=tempfile.mktemp(dir=dirPath)
 	try:
+		# TODO: turn into tryCopyFile?
```
Comment (michaelDCurran) [inline source/installer.py:580]: I don't quite follow what you mean about turning into tryCopyFile. I think that there are situations where you can rename a file but cannot write over it, such as a dll held by another application. Thus tryCopyFile would still leave behind the held file and the installer won't be able to write over it...

**Review by michaelDCurran — COMMENTED** [review]
At a minimum I think the psutil stuff is good. I wouldn't want the pr held up due to queries on changing the renames etc. We could do that  in beta and or master. But the psutil stuff is just a simple check and fail with a static list of exes. Very safe.

**Comment by CyrilleB79** [comment]
Which milestone did you want to use? 2024.3.4 milestone is clearly erroneous. #16122 has to be retagged too.

**Comment by cary-rowen** [comment]
It might be worth investigating #16123, which might also resolve #16122.
If #16123 does exist, it could also cause an unknown potential flaw.

File: source/installer.py [hunk 3]
```diff
@@ -2,12 +2,13 @@
...
 # This file is covered by the GNU General Public License.
 # See the file COPYING for more details.
-# Copyright (C) 2011-2023 NV Access Limited, Joseph Lee, Babbage B.V., Łukasz Golonka
+# Copyright (C) 2011-2024 NV Access Limited, Joseph Lee, Babbage B.V., Łukasz Golonka
 
 import ctypes
 import winreg
 import time
 import os
+import psutil
```
Comment (lukaszgo1) [inline source/installer.py:11]: This dependency should be included in our requirements file.

File: source/installer.py [hunk 4]
```diff
@@ -582,6 +590,7 @@ def tryRemoveFile(path,numRetries=6,retryInterval=0.5,rebootOK=False):
 			return
 		except OSError:
 			pass
+		log.debugWarning(f"Failed to delete file {tempPath}, attempt {count}/{numRetries}")
```
Comment (lukaszgo1) [inline source/installer.py:593]: Is there a reason why this is not in the `except` blog? If it would be moved there we could also log the exception which caused the deletion to fail.

File: source/installer.py [hunk 5]
```diff
@@ -633,7 +662,7 @@ def install(shouldCreateDesktopShortcut: bool = True, shouldRunAtLogon: bool = T
 	# so we shouldn't proceed.
 	# 2. The appropriate executable for nvda.exe will be determined by
 	# which executables exist after copying program files.
-	for f in ("nvda.exe","nvda_noUIAccess.exe","nvda_UIAccess.exe","nvda_service.exe","nvda_slave.exe"):
```
Comment (lukaszgo1) [inline source/installer.py:636]: There is a change of logic here - we no longer check for `nvda_service.exe`. While it is true NVDA no longer installs the service for some time now, it is possible, though unlikely, someone upgrades from the version where service was included. In general given this targets a patch release I would be pretty careful with introducing any logic changes, unless they are strictly required to fix the issue this PR intents to solve.

**Review by lukaszgo1 — CHANGES_REQUESTED** [review]

**Comment by Brian1Gaff** [comment]
I don't know if people are still having problems understanding how the portable  version is copied, but certainly when I started I did not realise that it would splurge its content into a root folder. Most  portable or installed products use a sensible safe default, creating the folder if its not present.
 Brian

-- 
***@***.*** 
Sent via blueyonder.(Virgin media) 
Please address personal E-mail to:-
***@***.***, putting 'Brian Gaff'
in the display name field.
  ----- Original Message ----- 
  From: Sean Budd 
  To: nvaccess/nvda 
  Cc: Subscribed 
  Sent: Tuesday, February 13, 2024 5:36 AM
  Subject: [nvaccess/nvda] Make installer process safer (PR #16174)


  Link to issue number:
  Fixes #16122

  Summary of the issue:
  NVDA process sometimes fails to quit correctly (see #16123).
  When running the installer, this causes the installer to fail partway through.
  If the installer fails partway through, the installation may end up in an unrecoverable state.

  For certain files, NVDA requests windows to move or delete these on reboot, if the move/delete fails initially.
  However, if the Windows API request fails, NVDA does not handle the failure the same way as it handles other failures - raising a retryable failure and reverting the change if possible.

  Description of user facing changes
  The NVDA installation process should be more resilient if the previous process is still running.

  Description of development approach
  NVDA uses psutil to check which process are running. If any known NVDA exe is running, the installation is aborted with a retry able failure. This means the user can attempt the installation again - i.e. waiting a bit longer for the process to hopefully exit.

  If NVDA fails to request windows to remove certain files on restart, this is handled the same as if the files are failed to be removed.

  More logging has been added and TODO comments have been raised over other potential code issues.

  Testing strategy:
    a..  Tested manually triggering a retriable failure to ensure the handling is as expected. 
    b..  Manual testing from the community is required to confirm the fix. 
  Known issues with pull request:
  The installation process should be atomic - i.e. if it fails, the whole operation should be reverted. This is not the case of the current design.

  There are many TODO comments which should be investigated if this PR does not solve the installer bug.
  If possible, these potential changes should probably be investigated in a subsequent release cycle with sufficient testing.

  Code Review Checklist:
    a..  Documentation: 
      a.. Change log entry 
      b.. User Documentation 
      c.. Developer / Technical Documentation 
      d.. Context sensitive help for GUI changes 
    b..  Testing: 
      a.. Unit tests 
      b.. System (end to end) tests 
      c.. Manual testing 
    c..  UX of all users considered: 
      a.. Speech 
      b.. Braille 
      c.. Low Vision 
      d.. Different web browsers 
      e.. Localization in other languages / culture than English 
    d..  API is compatible with existing add-ons. 
    e..  Security precautions taken. 

------------------------------------------------------------------------------

  You can view, comment on, or merge this pull request online at:
    https://github.com/nvaccess/nvda/pull/16174

  Commit Summary
    a.. 4a6717f Make installer process safer 
    b.. 0d77e8e update changes and build version 
    c.. 69bdcf5 update changes and build version 
    d.. f6a953b add todo 
  File Changes (4 files)
    a.. M appx/sconscript (4) 
    b.. M source/buildVersion.py (2) 
    c.. M source/installer.py (43) 
    d.. M user_docs/en/changes.t2t (7) 
  Patch Links:
    a.. https://github.com/nvaccess/nvda/pull/16174.patch 
    b.. https://github.com/nvaccess/nvda/pull/16174.diff

**Comment by seanbudd** [comment]
@Brian1Gaff - you are referring to #3113. This comment is totally off topic to this PR.

File: appx/sconscript [hunk 6]
```diff
@@ -63,7 +63,8 @@ signExec=env['signExec'] if env['certFile'] else None
 # Files from NVDA's distribution that cannot be included in the appx due to policy or security restrictions
 excludedDistFiles=[
 	'nvda_eoaProxy.exe',
```
[the remark this answers is not in the document]
Comment (seanbudd) [reply appx/sconscript:65]: I believe this has already been removed there, but nvda_service hasnt

**Comment by LeonarddeR** [comment]
I don't quite understand why we would have to roll out a patch release for this. While this issue was reported more frequently during the last few weeks, the current behavior has been there for almost a decade. This is neither a security fix nor a fix for a regression introduced in the 2023.3 cycle.

**Comment by michaelDCurran** [comment]
Agreed it is not a security fix. However, recently there have been a 
higher number of point releases, which has made the installer issue much 
more common.

My firm opinion is that either we put this into a beta and withhold any 
further patch releases to 2023.3, or we put it into a 2023.3 patch 
release, before any further patch releases.

Although the potential for this bug has existed for a long time, it has 
become more prevalent in the last few weeks. And it is at the point 
where we can't in good conscience make any further patch releases with 
out fixing it, as updates are leaving NVDA in a corrupt state for some.

**Comment by LeonarddeR** [comment]
> My firm opinion is that either we put this into a beta and withhold any further patch releases to 2023.3, or we put it into a 2023.3 patch release, before any further patch releases. 

I'd prefer the former. If the latter, I guess the fix can be applied as soon as a patch release is necessary (e.g. as part of a security fix)

**Comment by josephsl** [comment]
Hi, if it indeed involves introducing a new dependency (psutil, a dependency used by various add-ons), then I agree with Leonard that publishing beta 8 would be better. Thanks.

File: source/installer.py [hunk 7]
```diff
@@ -621,9 +629,27 @@ def tryCopyFile(sourceFilePath,destFilePath):
...
+	fullPathNvdaExes = {os.path.join(installDir, fn) for fn in _nvdaExes}
+	for process in psutil.process_iter():
+		if process.exe() in fullPathNvdaExes:
+			return True
+	return False
+
 
 def install(shouldCreateDesktopShortcut: bool = True, shouldRunAtLogon: bool = True):
 	prevInstallPath=getInstallPath(noDefault=True)
+	if _isNVDAAlive(prevInstallPath):
```
Comment (lukaszgo1) [inline source/installer.py:651]: This should handle  a case where NVDA was never installed on a given system:
```suggestion
	if prevInstallPath is not None and _isNVDAAlive(prevInstallPath):
```

File: source/installer.py [hunk 8]
```diff
@@ -633,7 +659,10 @@ def install(shouldCreateDesktopShortcut: bool = True, shouldRunAtLogon: bool = T
 	# so we shouldn't proceed.
 	# 2. The appropriate executable for nvda.exe will be determined by
 	# which executables exist after copying program files.
-	for f in ("nvda.exe","nvda_noUIAccess.exe","nvda_UIAccess.exe","nvda_service.exe","nvda_slave.exe"):
+	# Some exes are no longer used, but we remove them anyway from legacy copies.
+	# nvda_service.exe was removed in 2017.4 (#7625).
+	# nvda_eoaProxy.exe was removed in 2024.1 (#15544).
```
Comment (lukaszgo1) [inline source/installer.py:664]: I'm sorry for being painful here, but this, once again, changes the logic by removing EOA proxy way earlier compared to where it used to be removed. Note that it is possible for it to be still running at this point from the UAC screen which requested admin permissions for the installer. This will not only affect Windows 7, but also any more recent version of Windows which was upgraded in place from Win 7, and where NVDA was not reinstalled afterwards, as in that case the more recent support for  ease of access is not applied automatically. Long story short I'd suggest not to remove EOA proxy here, unless these scenarios are tested specifically, which IMHO is not worth the time.

**Review by lukaszgo1 — CHANGES_REQUESTED** [review]

**Comment by michaelDCurran** [comment]
> It is also unclear how this will help users to upgrade NVDA, granted 
it will not become corrupted, but if it fails to exit restarting the 
machine and attempting upgrade once again will cause users to end up in 
the same state in which old copy of NVDA is running preventing the upgrade.

You are correct that it will not necessarily assist the user in 
upgrading. However, we we really must avoid corrupting a user's existing 
install if at all possible.


 > I'd love to see #16123 investigated and will certainly support adding 
it into a patch release.

Absolutely. Any help would be appreciated here. Same with #16072, which 
I myself am still trying to investigate further.

File: source/installer.py (hunk 8 above)
Comment (seanbudd) [reply source/installer.py]: Note that Windows 7 is no longer in support, and 2023.3 is the last version with official support. However, I agree we can delay this change to 2024.1 easily

**Comment by seanbudd** [comment]
Please note that this bug has become a lot more common, and is a p1 issue causing installations to enter an unrecoverable state.

2024.1 is still a long way away. This fix will go through thorough rc, beta and alpha testing before a 2023.3.4 patch release comes out.

**Comment by hwf1324** [comment]
I can't test the issue because I haven't had an installation failure.

But I constructed the #16072 environment and tested some scenarios.

1. reproduce issue #16072. In previous versions of alpha-30888, nvda.exe remains on exit.
2. no problem installing NVDA via updated installer during NVDA runtime.
3. when installing updated NVDA via installer after exit, nvda.exe needs to be killed manually.
4. after installing this PR build and exiting NVDA, nvda.exe no longer seems to remain.

To summarize, seems to resolve #16072, but not sure if similar issues are resolved together, need more testing.

**Comment by CyrilleB79** [comment]
We are releasing more and more 2023.3.x releases. The last of them will remain as the last release compatible with Windows 7 and will be provided as such through a hard link on NV Access website. Thus it may be desirable to have the last changes in the change log also translated.

Since 2024.1 translation has already begun, could we imagine to ask exceptionally to translators a modified version of the change log for 2023.3.4 (or more) and merge them in a PR?

**Comment by hwf1324** [comment]
I overlooked an issue and am not sure if the version this PR is based on will have the #16072 issue.

File: source/installer.py [hunk 9]
```diff
@@ -622,9 +631,60 @@ def tryCopyFile(sourceFilePath,destFilePath):
...
+	tempDir = tempfile.mkdtemp()
+	for filepath in relativeFilepaths:
+		originalPath = os.path.join(installDir, filepath)
+		if not pathlib.Path(originalPath).exists():
+			log.debug(f"Skipping remove for non-existent file: {originalPath}")
+			continue
+		tempPath = os.path.join(tempDir, filepath)
+		pathlib.Path(tempPath).parent.mkdir(parents=True, exist_ok=True)
+		try:
+			os.rename(originalPath, tempPath)
```
Comment (michaelDCurran) [inline source/installer.py:673]: My manual testing shows that while NvDA is running, I can still rename nvda.exe with no issue. However, I cannot remove it. Therefore, this rename should be replaced with a copy and then remove.

**Review by michaelDCurran — CHANGES_REQUESTED** [review]

**Comment by seanbudd** [comment]
>The last of them will remain as the last release compatible with Windows 7 and will be provided as such through a hard link on NV Access website. 

The security fixes released in the subsequent patch releases have not been tested with Windows 7. NV Access is only recommending 2023.3 as the last compatible release with Wind0ws 7

**Comment by seanbudd** [comment]
@hwf1324 - this PR is meant to fix https://github.com/nvaccess/nvda/issues/16122 not #16072

**Comment by wmhn1872265132** [comment]
1. 2023.3.X can be installed manually on Windows 7.
2. After installing 2023.3.X, check for updates and the 2023.3 version will be detected, which seems to cause trouble to users.

Therefore, it is strongly recommended to allow upgrades to new patch versions. Although these updates may not be tested, in theory they should not cause a worse experience.

**Comment by seanbudd** [comment]
>Therefore, it is strongly recommended to allow upgrades to new patch versions. 

It is allowed, but not recommended. Windows 7 support was dropped as of 2023.3. Any patch releases after will not be tested with Windows 7.

>Although these updates may not be tested, in theory they should not cause a worse experience.

This is not true - the changes to HTML browsable message may break a large number of dialogs on Windows 7

**Comment by lukaszgo1** [comment]
@seanbudd  This approach to dropping Windows 7 support is quite unusual. Given that 5% of your users are running it, the basic thing to do would be to either add the link to 2023.0 to NV Access's downloads web page (as soon as the first untested patch release was published), and clearly state that it should be used on Windows 7, or test these patch releases. While I understand that you have limited resources and performing this testing yourselves may be difficult, one of NVDA's strengths it its community, as far as I can tell you didn't reach out to us asking for testing. Also while at it can we please have a clarification of these patch releases and Windows Server  2012 which is still under extended support until 2026?

**Comment by josephsl** [comment]
Hi, Server 2012/2012 R2 are receiving extended security updates, and recently Python’s PEP 11 was edited to state that future Python releases will support Windows releases under extended support (as of now, Windows 10 Version 1607/Server 2016 in Python 3.13 onwards). Thanks.

**Comment by codeofdusk** [comment]
I thought Windows 7 support was dropped as of 2024.1? (per the changelog).

Any patch releases on the 2023.x branch should support 7, per my intuition. Any subsequent full releases (2024.1 onwards) should not.

**Comment by CyrilleB79** [comment]
To say things more clearly:
The change logs of NVDA 2023.3.1, 2023.3.2 and 2023.3.3 do not indicate that Windows 7 support has been dropped. If it should be the case, it's at least a clear issue in the communication of this information.

For now (in its beta stage), according to its change log, the first version that drops Windows 7 compatibility is 2024.1.

My recommendation would be to do as advertised largely in the change log and elsewhere in the community, i.e. 2024.1 to be the first version dropping Windows 7 support.
If it's not technically possible, I would ask that a much wider and clearer communication be made on that matter. Thanks.
Cc @gerald-hartig

**Comment by seanbudd** [comment]
Windows Server 2012 are still supported by Python, however we are not actively testing or fixing these bugs.

I think there has been some confusion around NV Access development support and whether or not something runs.
2024.1 will not run on Windows 7 and 8. In 2023 NV Access stopped active development support for Windows 7. 
We are avoiding encouraging updates to 2023.3.X patch releases on Windows 7 as these are under tested and may be worse compared to 2023.3. Windows 7 users have nothing to gain, and potential risks, for updating from 2023.3

**Comment by cary-rowen** [comment]
But there is no way to prevent Windows 7 users from manually installing 2023.3.X. As far as the current publicity is concerned, 2024.1 is the version that clearly does not support Windows 7.

It may be natural for people to think that Windows 7 should choose the last version allowed to run on Windows 7. Few people know that NV Access has No longer recommended this.

**Comment by seanbudd** [comment]
@cary-rowen - the release posts advertise this information.
The reality is Windows 7 has had no active development support in NVDA for sometime, we are not interested in spending additional developer time or user support for Windows 7 users.  All NV Access plans to do as of 2024 is advise on which version we recommend for Windows 7 users, which we are trying to cover in all relevant locations on our website.

**Review by michaelDCurran — DISMISSED** [review]
